File: tunnel/ts/src/wire.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { at, mount, pipe, type Endpoint } from '@nightseam/duplex';
import { DuplexPeer, callWire, createDispatcher, handleWire } from '@nightseam/runtime';
import { Tunnel } from './index.ts';

async function tunnels() {
  const [a, b] = pipe();
  const left = new DuplexPeer(),
    right = new DuplexPeer({ role: 'server' });
  await Promise.all([left.attach(a), right.attach(b)]);
  return {
    client: new Tunnel(left),
    server: new Tunnel(right),
    left,
    right,
    close: () => {
      left.close();
      right.close();
    },
  };
}

test('channel is a prepared wire before its first selection or mounted use', async (t) => {
  const pair = await tunnels();
  t.after(pair.close);
  let opens = 0;
  const observer = {
    observe: (event: { type: string }) => {
      if (event.type === 'connection.opened') opens++;
    },
  };
  const digest = 'a'.repeat(64);
  const opened: Endpoint & { id: number; digest: string } = await pair.client.open('wire', digest, { observer });
  const pending = callWire(opened, ['deep', 'echo'], 'ok');
  await Promise.resolve();
  const accepted = await pair.server.accept({
    observer,
    prepare: (peer) => {
      const dispatcher = createDispatcher(peer.wire());
      t.after(() => dispatcher.close());
      handleWire(dispatcher, ['deep', 'echo'], (value) => value);
    },
  });
  assert.equal(await pending, 'ok');
  assert.equal(opened.digest, digest);
  assert.equal(accepted.digest, digest);
  assert.equal(opens, 2);
  const selected = at(mount(new Map([['route', opened]])), ['route', 'deep']);
  assert.equal(await callWire(selected, ['echo'], 'selected'), 'selected');
  assert.equal(await pair.client.channel(opened.id, { prepare: () => assert.fail('lookup rebuilt peer') }), opened);
  assert.equal(opens, 2);
  assert.equal(pair.client.connection(opened.id), undefined);
  accepted.close();
  assert.equal(pair.left.status, 'connected');
  assert.equal(pair.right.status, 'connected');
});

test('raw connection refuses a second wire reader and retains its identity', async (t) => {
  const pair = await tunnels();
  t.after(pair.close);
  const raw = await pair.client.openConnection('raw', '');
  await assert.rejects(pair.client.channel(raw.id), { code: 'channel_invalid' });
  assert.equal(pair.client.connection(raw.id), raw);
});

test('peer preparation runs after validation and cleans registered wire receivers on failure', () => {
  let prepared = 0,
    closed = 0;
  assert.throws(
    () =>
      new DuplexPeer({
        queueCapacity: 0,
        prepare: () => {
          prepared++;
        },
      }),
  );
  assert.equal(prepared, 0);
  assert.throws(
    () =>
      new DuplexPeer({
        prepare: (peer) => {
          prepared++;
          peer.wire().receive({
            closed: () => {
              closed++;
            },
            message: () => {},
          });
          throw new Error('prepare failed');
        },
      }),
    /prepare failed/,
  );
  assert.equal(prepared, 1);
  assert.equal(closed, 1);
  assert.throws(() => new DuplexPeer({ prepare: async () => {} }), {
    code: 'invalid_options',
    message: 'prepare must complete synchronously.',
  });
});

test('server-opened wire channel preserves channel parity and carries calls in both directions', async (t) => {
  const pair = await tunnels();
  t.after(pair.close);
  const options = {
    prepare: (peer: DuplexPeer) => {
      const dispatcher = createDispatcher(peer.wire());
      t.after(() => dispatcher.close());
      handleWire(dispatcher, ['echo'], (value) => value);
    },
  };
  const opened = await pair.server.open('reverse', '', options),
    accepted = await pair.client.accept(options);
  assert.equal(opened.id % 2, 0);
  for (const wire of [opened, accepted]) assert.equal(await callWire(wire, ['echo'], 'both ways'), 'both ways');
});

File: tunnel/go/wire_test.go
package tunnel_test

import (
	"context"
	"encoding/json"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"strings"
	"sync/atomic"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
	"github.com/Bitspark/nightseam/runtime/go"
	"github.com/Bitspark/nightseam/tunnel/go"
)

type wireOpens struct{ count atomic.Int32 }

func (w *wireOpens) Observe(event runtime.ObserverEvent) {
	if _, ok := event.(runtime.ConnectionOpened); ok {
		w.count.Add(1)
	}
}

func TestChannelIsAPreparedWireBeforeSelection(t *testing.T) {
	client, server := peers(t, tunnel.Options{})
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	observer := &wireOpens{}
	acquired, stopAcquisition := context.WithCancel(ctx)
	digest := strings.Repeat("a", 64)
	opened, err := client.Open(acquired, "wire", digest, runtime.Options{Observer: observer})
	if err != nil {
		t.Fatal(err)
	}
	stopAcquisition()
	// The first request can already be waiting in the raw channel when accepted.
	result := make(chan error, 1)
	go func() {
		var got string
		err := runtime.CallWire(ctx, opened, []string{"deep", "echo"}, "ok", &got)
		if err == nil && got != "ok" {
			t.Errorf("got %q", got)
		}
		result <- err
	}()
	accepted, err := server.Accept(ctx, runtime.Options{Observer: observer, Prepare: func(peer *runtime.Peer) error {
		binding, err := runtime.NewDispatcher(peer.Wire())
		if err != nil {
			return err
		}
		_, err = runtime.HandleWire(binding, []string{"deep", "echo"}, func(_ context.Context, raw json.RawMessage) (any, error) { return raw, nil })
		return err
	}})
	if err != nil {
		t.Fatal(err)
	}
	if err := <-result; err != nil {
		t.Fatal(err)
	}
	if opened.Digest != digest || accepted.Digest != digest {
		t.Fatal("prepared channel lost declaration digest")
	}
	var _ bitwire.Wire = (*tunnel.Channel)(nil)
	if observer.count.Load() != 2 {
		t.Fatalf("construction peers %d", observer.count.Load())
	}
	selected := duplex.At(duplex.Mount(map[string]bitwire.Endpoint{"route": opened}), []string{"route", "deep"})
	var got string
	if err := runtime.CallWire(ctx, selected, []string{"echo"}, "selected", &got); err != nil || got != "selected" {
		t.Fatalf("selected %q %v", got, err)
	}
	cached, ok, err := client.Channel(opened.ID, runtime.Options{Prepare: func(*runtime.Peer) error { t.Fatal("lookup rebuilt peer"); return nil }})
	if err != nil || !ok || cached != opened || observer.count.Load() != 2 {
		t.Fatal("selection or lookup allocated a peer", err)
	}
	if _, ok := client.Connection(opened.ID); ok {
		t.Fatal("wire channel also handed out raw reader")
	}
	if err := accepted.Close(duplex.CodeNormal, "done"); err != nil {
		t.Fatal(err)
	}
	if client.Peer().Err() != nil || server.Peer().Err() != nil {
		t.Fatal("channel close ended outer peer")
	}
}

func TestRawConnectionCannotAcquireASecondWireReader(t *testing.T) {
	client, _ := peers(t, tunnel.Options{})
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	raw, err := client.OpenConnection(ctx, "raw", "")
	if err != nil {
		t.Fatal(err)
	}
	if _, ok, err := client.Channel(raw.ID, runtime.Options{}); !ok || err == nil {
		t.Fatalf("raw claimed as wire: %t %v", ok, err)
	}
	if got, ok := client.Connection(raw.ID); !ok || got != raw {
		t.Fatal("raw lookup identity changed")
	}
}

func TestServerOpenedWireChannelUsesOnePreparedPeerOnEachSide(t *testing.T) {
	client, server := peers(t, tunnel.Options{})
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	prepare := runtime.Options{Prepare: func(peer *runtime.Peer) error {
		binding, err := runtime.NewDispatcher(peer.Wire())
		if err != nil {
			return err
		}
		_, err = runtime.HandleWire(binding, []string{"echo"}, func(_ context.Context, raw json.RawMessage) (any, error) { return raw, nil })
		return err
	}}
	opened, err := server.Open(ctx, "reverse", "", prepare)
	if err != nil {
		t.Fatal(err)
	}
	accepted, err := client.Accept(ctx, prepare)
	if err != nil {
		t.Fatal(err)
	}
	if opened.ID%2 != 0 {
		t.Fatal("server changed channel parity")
	}
	for _, wire := range []bitwire.Wire{opened, accepted} {
		var result string
		if err := runtime.CallWire(ctx, wire, []string{"echo"}, "both ways", &result); err != nil || result != "both ways" {
			t.Fatalf("call %q %v", result, err)
		}
	}
}
